### PR TITLE
Add principal as argument to role-assignments

### PR DESCRIPTION
**Pull request by HonzaHrazdira**
##### SUMMARY

Adds `principal` argument to role-assignments module. 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`fusion_ra`

##### ADDITIONAL INFORMATION
Principal is an ID of either User or API client. `principal` argument can be used instead of `user` argument which is a human readable identifier that in current implementation works only for Users not API Clients.

### CODE REVIEW

File: playbooks/simple/create_role_assignment.yaml
Comment (andreepyro) [inline playbooks/simple/create_role_assignment.yaml:23]: Why this commented lines?

File: tests/functional/test_fusion_ra.py
Comment (andreepyro) [inline tests/functional/test_fusion_ra.py:214]: If parameters are wrong `list_role_assignments` shouldn't be called at all. Do the test pass?

**Review by andreepyro — CHANGES_REQUESTED** [review]
LGTM. Please add change fragment (see `/changelogs` folder). Also, don't forget to add the example playbook to  [example-playbooks](https://github.com/PureStorage-OpenConnect/ansible-playbook-examples/tree/master/fusion) repository.

**Comment by andreepyro** [comment]
One more thing, please fix the linting errors.

File: playbooks/simple/create_role_assignment.yaml
Comment (HonzaHrazdira) [reply playbooks/simple/create_role_assignment.yaml:23]: whoops, I created `[create_role_assignment.yaml](https://github.com/Pure-Storage-Ansible/Fusion-Collection/pull/135/files#diff-df40034ca1f34f237c3f4055e887a546d99282c4c15277367c7a949a8d0a0c87)` just so I can test the changes. Should I delete the file or uncomment the tasks and leave it there together with other simple playbooks?

Comment (andreepyro) [reply playbooks/simple/create_role_assignment.yaml:23]: It's up to you. If there's no reason why you added the playbook, you can just delete it.

File: plugins/modules/fusion_ra.py
Comment (jankodera-ps) [inline plugins/modules/fusion_ra.py:101]: The original idea was that the user field was supposed to be replaced by a principal field, similar to the API and have it just accept users as well for convenience, which is a bit different than this implementation which treats it as a separate field which I think will be confusing to the SDK users. They should not collide as user name is a human readable string and generated principal IDs are UUID. But don't know if you guys did not make some other decision.

**Comment by jankodera-ps** [comment]
The CLA is required even for us

File: plugins/modules/fusion_ra.py
Comment (HonzaHrazdira) [reply plugins/modules/fusion_ra.py:101]: We discussed this with @genalt and @LukasZahradnik and the result was that we should keep the `user` field for backward compatibility and align with our other SDKs by adding the `principal`. But I kinda agree with you that it might be more convenient for the user. @genalt and @LukasZahradnik WDYT?

File: tests/functional/test_fusion_ra.py
Comment (HonzaHrazdira) [reply tests/functional/test_fusion_ra.py]: fixed in [d9f52e1](https://github.com/Pure-Storage-Ansible/Fusion-Collection/pull/135/commits/d9f52e1f74613e780884dbedee1eac16edea0e29) and [84a7043](https://github.com/Pure-Storage-Ansible/Fusion-Collection/pull/135/commits/84a70432496f90bad925c8f4c14bc777a7aaa399)

File: plugins/modules/fusion_ra.py
Comment (andreepyro) [reply plugins/modules/fusion_ra.py:101]: If it's for backward compatibility only, you can deprecate the old field so we can remove it in 2.0.0

Comment (HonzaHrazdira) [reply plugins/modules/fusion_ra.py:101]: Also there is another story to support also `api-client` field to align with `hmctl role-assignment list --api-client`. It might become difficult to find out whether the input to principal field is user name or api-client name

File: plugins/modules/fusion_ra.py
Comment (andreepyro) [inline plugins/modules/fusion_ra.py:151]: I know it's not your code, but this PR seems like a good place to address this. Can you please replace `for assign in range(0, len(assignments))` by `for assign in assignments`? It's much simpler and removes extra complexity :D

File: plugins/modules/fusion_ra.py
Comment (LukasZahradnik) [reply plugins/modules/fusion_ra.py:101]: @jankodera-ps We decided to have both `user` and `principal` fields. It might be more confusing that the `principal` supports both arbitrary user name and generated principal ID. On top of that, you have to distinguish between those. Then there is also API Client name, which users might assume `principal` also supports..
@andreepyro We want to keep the `user` field.

Comment (LukasZahradnik) [reply plugins/modules/fusion_ra.py:101]: @jankodera-ps The idea was that all you need in the TF provider is the `principal` field, because users can provide the principal ID by using references (e.g., `fusion_api_client.my_client.id`)

Comment (jankodera-ps) [reply plugins/modules/fusion_ra.py:101]: @LukasZahradnik sure, principal is actually in the api, user/api_client is convenience

File: plugins/modules/fusion_ra.py
Comment (HonzaHrazdira) [reply plugins/modules/fusion_ra.py]: Sure, good call. Fixed in [fc973ce](https://github.com/Pure-Storage-Ansible/Fusion-Collection/pull/135/commits/fc973cea2dac05fa3d8f761c50b6695b3db86675).

**Comment by HonzaHrazdira** [comment]
> The last linting error seems to be caused by a [file outside of this PR](https://github.com/Pure-Storage-Ansible/Fusion-Collection/blob/master/tests/sanity/ignore-2.16.txt). I don't think it is actually triggered by this PR, or is it? @sdodsley could you take a look, please?

I was able to fix the ansible-lint by changing the version in workflow to `v6.11.0` (latest is `v6.16.0`). It looks like newer versions don't allow to ignore this "return-syntax-error". 

The error msg from linter:`Ignore file contains validate-modules:return-syntax-error at line 1, which is not a permitted ignore`

**Comment by sdodsley** [comment]
Let me speak with the ansible core team to see what has changed

**Comment by andreepyro** [comment]
I've just introduced a PR fixing this https://github.com/Pure-Storage-Ansible/Fusion-Collection/pull/136, please review.

**Review by sdodsley — APPROVED** [review]

**Comment by sdodsley** [comment]
#137 solves all the issues and allows us to keep using the correct `ansible-lint` version, ie use `main`

**Review by jankodera-ps — APPROVED** [review]

**Review by DenDen039 — COMMENTED** [review]
LGTM, but you need to add changelog fragment

**Review by andreepyro — DISMISSED** [review]
LGTM. Once @DenDen039 approves this PR, we can merge this as others already approved.

edit: please fix the lint error :)

**Review by andreepyro — APPROVED** [review]

**Pull request merged**
